pmda/parallel.py

```diff
@@ -307,144 +367,123 @@ def run(self,
                 # explicit instead of 'for ts in u.trajectory[bslice]'
                 # so that we can get accurate timing.
-                ts = u.trajectory[i]
+                self._ts = self._trajectory[i]
             # record compute time per frame
             with timeit() as b_compute:
-                res = self._reduce(res, self._single_frame(ts, agroups))
+                res = self._reduce(res, self._single_frame())
             times_io.append(b_io.elapsed)
             times_compute.append(b_compute.elapsed)
 
         # calculate io and compute time per block
         return np.asarray(res), np.asarray(times_io), np.asarray(
-            times_compute), b_universe.elapsed, wait_end, np.sum(
+            times_compute), wait_end, np.sum(
             times_io), np.sum(times_compute)
 
     @staticmethod
     def _reduce(res, result_single_frame):
         """ 'append' action for a time series"""
         res.append(result_single_frame)
         return res
+
+    def __getstate__(self):
+        #  To make sure Universe is pickled before Atomgroup,
+        base_dict = self.__dict__
+        universe_dict = {}
+        for key, item in base_dict.items():
+            if(isinstance(item, mda.Universe)):
+                universe_dict[key] = item
+        universe_dict.update(base_dict)
+        return universe_dict
+
+    #  dask-related functions
+
+    def __dask_graph__(self):
+        return self._dsk
+
+    def __dask_keys__(self):
+        return self._keys
+
+    #  it uses multiprocessing scheduler in default
+    __dask_scheduler__ = staticmethod(dask.multiprocessing.get)
+
+    def __dask_postcompute__(self):
+        return self._post_reduce, ()
+
+    def _post_reduce(self, res):
+        # hack to handle n_frames == 0 in this framework
+        if len(res) == 0:
+            # everything else wants list of block tuples
+            res = [([], [], [], self.wait_start, 0, 0)]
+        # record conclude time
+        with timeit() as conclude:
+            self._results = np.asarray([el[0] for el in res])
+            # save the frame numbers for all blocks
+            self._conclude()
+        # put all time information into the timing object
+        self.timing = Timing(
+            np.hstack([el[1] for el in res]),
+            np.hstack([el[2] for el in res]), 0,
+            self.time_prepare,
+            self.time_prepare_dask,
+            conclude.elapsed,
+            # waiting time = wait_end - wait_start
+            np.array([el[3] - self.wait_start for el in res]),
+            np.array([el[4] for el in res]),
+            np.array([el[5] for el in res]))
+
+        #  this is crucial if the analysis does not iterate over
+        #  the whole trajectory.
+        self._trajectory.rewind()
+        self._dsk = {}
+        self._keys = []
+        self._job_prepared = False
+
+    def __dask_postpersist__(self):
+        #  we don't need persist implementation.
+        raise NotImplementedError
```

Comment:
Will it not be possible to `persist`? 

Presumably, that would have been possible previously if we had chosen `persist` in `run` instead of `compute()`.